commons/src/main/java/org/infinispan/commons/collections/MiniList.java

```diff
@@ -0,0 +1,298 @@
+            case SINGLE:
+                return equal(holder, o) ? 0 : -1;
+            case ARRAY:
+                Object[] array = (Object[]) holder;
+                for (int i = size - 1; i >= 0; --i) {
+                    if (equal(array[i], o)) return i;
+                }
+                return -1;
+        }
+        throw new IllegalStateException();
+    }
+
+    @Override
+    public void clear() {
+        modCount++;
+        size = 0;
+        holder = null;
+        type = Type.EMPTY;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        switch (type) {
+            case EMPTY:
+                return false;
+            case SINGLE:
+                return equal(holder, o);
+            case ARRAY:
+                Object[] array = (Object[]) holder;
+                for (int i = 0; i < size; ++i) {
+                    if (equal(array[i], o)) return true;
+                }
+                return false;
+        }
+        throw new IllegalStateException();
+    }
+
+    @Override
+    public T remove(int index) {
+        checkBounds(index);
+        modCount++;
+        T tmp;
+        switch (type) {
+            case SINGLE:
+                tmp = (T) holder;
+                holder = null;
+                size = 0;
+                type = Type.EMPTY;
+                return tmp;
+            case ARRAY:
+                Object[] array = (Object[]) holder;
+                tmp = (T) array[index];
+                System.arraycopy(array, index + 1, array, index, size - index - 1);
+                array[--size] = null; // for GC
+                return tmp;
+        }
+        throw new IllegalStateException();
+    }
+
+    @Override
+    public void add(int index, T element) {
+        checkBoundsForAdd(index);
+        modCount++;
+        Object[] array;
+        switch (type) {
+            case EMPTY:
+                holder = element;
+                type = Type.SINGLE;
+                size = 1;
+                return;
+            case SINGLE:
+                array = new Object[INIT_ARRAY_SIZE];
+                if (index == 0) {
+                    array[0] = element;
+                    array[1] = holder;
+                } else {
+                    array[0] = holder;
+                    array[1] = element;
+                }
+                holder = array;
+                size = 2;
+                type = Type.ARRAY;
+                return;
+            case ARRAY:
+                array = ensureCapacity(size + 1);
+                System.arraycopy(array, index, array, index + 1, size - index);
+                array[index] = element;
+                size++;
+                return;
+        }
+        throw new IllegalStateException();
+    }
+
+    private Object[] ensureCapacity(int capacity) {
+        Object[] array = (Object[]) holder;
+        if (capacity > array.length) {
+            array = Arrays.copyOf(array, Math.max(capacity, array.length * 2));
```

Comment:
I'll leave changing 'x * 2' to 'x << 1' on the compiler.
Regarding integer overflows, you're right... I was not really expecting to store 1G entries in this list.